### PR TITLE
Improvements to handling and add constant passed to dot operator

### PR DESCRIPTION
**Pull request by pfultz2**
* This will rewrite dot operators like `X(Y + b)` to `XY + Xb` when `b` is constant as we can fold the add away.
* This improves handling pointwise with broadcasted operators, this helps improves const propagation.
* Improve gemm fusion with a mul_add
* Improve support for broadcast shapes in gemm

### CODE REVIEW

**Comment by pfultz2** [comment]
This needs tests.

**Comment by umangyadav** [comment]
> This needs tests.

We should add unittests.

**Comment by umangyadav** [comment]
> * This will rewrite dot operators like `X(Y + b)` to `XY + Xb` when `b` is constant as we can fold the add away.
X is also constant and then, `Xb` can be folded.

File: src/simplify_algebra.cpp [hunk 1]
```diff
@@ -267,28 +303,26 @@ struct find_double_add_lit_broadcast
...
-        auto x_ins = r.instructions["x"];
-        auto y_ins = r.instructions["y"];
-
-        auto xbroadcast = any_cast<op::broadcast>(x_ins->get_operator());
-        auto ybroadcast = any_cast<op::broadcast>(y_ins->get_operator());
-
-        if(xbroadcast.axis != ybroadcast.axis)
+        auto ins        = r.result;
+        auto broadcasts = ins->inputs();
+        if(broadcasts.empty())
```
[the remark this answers is not in the document]
Comment (pfultz2) [reply src/simplify_algebra.cpp:312]: Well if it is a nullary operator, which is unlikely but there for extra error checking.

**Comment by pfultz2** [comment]
Unit tests added.

**Review by umangyadav — APPROVED** [review]

**Review by turneram — APPROVED** [review]

**Pull request merged**